### PR TITLE
[ADD] Added new module web_widget_prefixed_url

### PR DESCRIPTION
**Pull request by Swapnesh-SerpentCS**
A simple module for adding 'callto:' widget to your odoo fields

`<field name="skype_name" widget="callto"/>`

Ps: I will update PR with README.rst and other fixes If this module is usefull for Community and will be accepted here.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Thanks for this. Please call it `web_widget_callto` instead for more homogeneous name.

File: web_phone_widget_callto/LICENSE/LICENSE
Comment (pedrobaeza) [inline web_phone_widget_callto/LICENSE/LICENSE]: LICENSE is put only on root folder, and this one seems outdated.

File: web_phone_widget_callto/__manifest__.py
Comment (sudhir-erpharbor) [inline web_phone_widget_callto/__manifest__.py]: website should be "https://github.com/OCA/web"

**Comment by sudhir-erpharbor** [comment]
Please add README files

**Comment by Swapnesh-SerpentCS** [comment]
@sudhir-erpharbor  Thanks for review but PR is still in the Draft stage.
The idea was to work on this only if this Module's feature is helpful for the community and will be accepted here.

@pedrobaeza  Will rename module and adept it according to OCA standards.

**Comment by Swapnesh-SerpentCS** [comment]
> Please add README files

@pedrobaeza  
Do we need all README files for v11 too? (I think it is from v12 only)

**Comment by pedrobaeza** [comment]
The requirement is to have a README.rst file in the module. README by fragments is only a helper for generating a proper README.rst easily, so it's recommended, but not mandatory. And it serves for all the versions of Odoo/OpenERP (even 6.1 if you want).

**Comment by Swapnesh-SerpentCS** [comment]
@pedrobaeza @sudhir-erpharbor This can be reviewed now.

**Comment by sudhir-erpharbor** [comment]
@Swapnesh-SerpentCS 
How does it work? I tried it on runbot by adding the widget. When I click on the skype name I entered in the field, it just opens a new tab with this URL: "callto:live:sudhir_362"

Correct me if I am doing anything wrong.

**Steps I followed on runbot:**
I used the email field in the partner and added callto widget (I removed the email widget from the field) and then I entered the skype name.

**Comment by Swapnesh-SerpentCS** [comment]
I just logged in to Test this on Runbot and Someone changed Login details.
Screenshot (18)

>  it just opens a new tab with this URL: "callto:live:sudhir_362"

It will open the External App (Skype) if supported.
Screenshot (19)


Ps: `skype:` will be used to Initiate Skype call directly from the browser. (skype:sudhir_362)

**Comment by sudhir-erpharbor** [comment]
Thanks for your reply. It is not working for me though. I think it has to do something with the version of the browser. Did you check that on other browsers as well?

**Comment by Swapnesh-SerpentCS** [comment]
> Thanks for your reply. It is not working for me though. I think it has to do something with the version of the browser. Did you check that on other browsers as well?

For `callto:` you may need VOIP application (or Skype Phone call) to installed on machine, I didn't checked this on my machine yet but I think it should works.
(`skype:skype_name` works well form me. (In Ubuntu too) )


Ps: I think it should be `callto://skype_phone`  (Instaed of `callto:skype_phone`) according to https://mycyberuniverse.com/create-html-link-that-starts-skype-call.html.
Can anoyone confirm ? (Or I will update README accordingly)

**Review by tarteo — COMMENTED** [review]
Why not use `web.basic_fields.FieldPhone`?

**Comment by Swapnesh-SerpentCS** [comment]
or Maybe `FieldEmail` ? (`FieldPhone` extends `FieldEmail` )

**Comment by tarteo** [comment]
Yea, but isn't this module doing the exact same thing as `web.basic_fields.FieldPhone`? What's the difference

**Comment by Swapnesh-SerpentCS** [comment]
Prefix is different, Few Apps supports `callto:`

**Review by tarteo — CHANGES_REQUESTED** [review]
Ok, then it's better to use `web.basic_fields.FieldEmail` or `web.basic_fields.FieldPhone`

**Comment by Swapnesh-SerpentCS** [comment]
> Ok, then it's better to use `web.basic_fields.FieldEmail` or `web.basic_fields.FieldPhone`

Could you please Review again? (Not sure If this is what you suggested to do)

File: web_widget_callto/static/src/js/widget_callto.js
Comment (tarteo) [inline web_widget_callto/static/src/js/widget_callto.js]: `init`, `getFocusableElement`, `_renderReadonly` and `supportedFieldTypes` are not needed anymore.

Comment (Swapnesh-SerpentCS) [reply web_widget_callto/static/src/js/widget_callto.js]: True, (As we are using `FieldEmail` now)
Fixed and Pushed.

**Comment by Swapnesh-SerpentCS** [comment]
@pedrobaeza  @tarteo 
How can I trigger Travis build again?

_EDIT: It's Done now, I don't know why It was pending since long._ (https://travis-ci.org/OCA/web/builds/515119882)

**Comment by Swapnesh-SerpentCS** [comment]
Instead of Individual module like `web_widget_callto` / `web_widget_tel` / `web_widget_skype` Can we have something like `web_widget_prefix` where we can add dynamic Prefix ?


Eg, `<field name="skype_name" widget="perfix" options="{'perfix_name': 'callto'}"/>`

I haven't worked on this but seems more generic to me.

What are your opinion ? @JayVora-SerpentCS @pedrobaeza @tarteo

**Comment by tarteo** [comment]
@Swapnesh-SerpentCS Yes sounds good. :+1:

**Comment by pedrobaeza** [comment]
About generic module name, I would go for `web_widget_url_prefix`. About Travis rebuild, just make an extra commit, rebase, or ask us for triggering again the build. Travis is very slow lately though.

**Comment by Swapnesh-SerpentCS** [comment]
@pedrobaeza  It might be confusing with existing `widget="url" `

**Comment by pedrobaeza** [comment]
Well, I don't think so, but if you prefer, use `web_widget_prefixed_url`, and the widget name `prefixed_url` in contrast to `url`.

**Comment by Swapnesh-SerpentCS** [comment]
Hello @tarteo @pedrobaeza 

Added a new way to manage this better.

Eg `<field name="skype_name" widget="prefixed_url" options="{'prefix_name': 'skype'}"/>
`

**Comment by Swapnesh-SerpentCS** [comment]
Any reviews please  ? (Apart from LINT_CHECK)

**Comment by Swapnesh-SerpentCS** [comment]
Someone here for Review ?

**Review by legalsylvain — APPROVED** [review]
LGTM code review. No test.

Thanks !

File: web_widget_prefixed_url/static/src/js/widget_prefixed_url.js
Comment (Tardo) [inline web_widget_prefixed_url/static/src/js/widget_prefixed_url.js]:
```suggestion
            if (!_.isEmpty(this.attrs.options.prefix_name)) {
```

**Review by Tardo — CHANGES_REQUESTED** [review]

**Comment by simahawk** [comment]
Agree on general remarks already posted.
About generic module name: can't we merge this into https://github.com/OCA/web/tree/11.0/web_widget_url_advanced ?

**Comment by Swapnesh-SerpentCS** [comment]
> Agree on general remarks already posted.
> About generic module name: can't we merge this into https://github.com/OCA/web/tree/11.0/web_widget_url_advanced ?

That will makes confusion and both refer to different functionalities.

`web_widget_prefixed_url` just adds dynamic `prefix`  (Indeed It will be more useful to use along with `web_widget_url_advanced` that module)

Ps: I will take a look if there are any conflicts using both modules or not (Just to be sure).

**Comment by simahawk** [comment]
@Swapnesh-SerpentCS you are adding a new widget, that one is improving the url widget.
All in all we are talking about enhanced features for links (being pure URLs or emails).
I think they can coexist. Also, having to install/maintain/migrate/etc only 1 module seems better. :wink:

**Comment by Swapnesh-SerpentCS** [comment]
That's also True. :+1: 
What do you think about this  @pedrobaeza @tarteo?

**Comment by pedrobaeza** [comment]
Merging both modules seems a good idea for me too, but I understand that it's more work now in this point, so proceed as you prefer, indicating at least in the roadmap about the merging if you don't do it.

**Comment by Swapnesh-SerpentCS** [comment]
`web_widget_url_advanced` module defines field as

`<field name="some_url" widget="url" text_field="another_field" />`


Here we can add one more option `options="{'perfix_name': 'callto'}"` or `perfix_name="callto"`

Can used as
`<field name="some_url" widget="url" perfix_name="callto" />`
or 
`<field name="some_url" widget="url" text_field="another_field"   perfix_name="callto" />`


I will prefer adding `perfix_name="callto"` instaed of option. 

Ps: Adding possible fidd when we are using `options="{'perfix_name': 'callto'}"` 
```diff
diff --git a/web_widget_url_advanced/static/src/js/url_widget.js b/web_widget_url_advanced/static/src/js/url_widget.js
index 51e38e44..805f0eda 100644
--- a/web_widget_url_advanced/static/src/js/url_widget.js
+++ b/web_widget_url_advanced/static/src/js/url_widget.js
@@ -24,6 +24,26 @@ odoo.define('web_widget_url_advanced', function (require) {
                 this.attrs.text = field_value;
             }
         },
+
+        /**
+         * In readonly, emails should be a mailto: link with proper formatting.
+         *
+         * @override
+         * @private
+         */
+        _renderReadonly: function () {
+            if (_.isEmpty(this.attrs.options.prefix_name)) {
+                this.$el.text(this.attrs.text || this.value)
+                    .addClass('o_form_uri o_text_overflow')
+                    .attr('target', '_blank')
+                    .attr('href', this.value);
+            } else {
+                var prefix = this.attrs.options.prefix_name;
+                this.$el.text(this.attrs.text || this.value)
+                    .addClass('o_form_uri o_text_overflow')
+                    .attr('href', prefix + ':' + this.value);
+            }
+        },
     });
 
 });
```

**Comment by Swapnesh-SerpentCS** [comment]
For now, we can merge this PR as efforts have been already made and for v12 we can add this functionality on `web_widget_url_advanced` module.

**Comment by Swapnesh-SerpentCS** [comment]
@pedrobaeza  Can we go ahead with for now?
(and will add this feature in `web_widget_url_advanced` for v12?)

**Comment by pedrobaeza** [comment]
@astirpe is the one with requested changes review, so he must unblock this.

**Comment by astirpe** [comment]
I think @tarteo and @Tardo were the ones reviewing this PR.

**Comment by pedrobaeza** [comment]
Right, sorry, I have mixed both.

**Comment by Swapnesh-SerpentCS** [comment]
@tarteo Could you please have a look again? (As you have Reviewed this in the Past)

**Comment by Swapnesh-SerpentCS** [comment]
@pedrobaeza  @tarteo  Hope to get merged today. :smiley:

**Comment by pedrobaeza** [comment]
Are you here in the sprint, @Swapnesh-SerpentCS ?

**Comment by Swapnesh-SerpentCS** [comment]
> Are you here in the sprint ?

Nope, :disappointed:  Still waiting for Passport to arrive. (I hope I get it by today so I can attend Odoo days at least)

**Comment by pedrobaeza** [comment]
Oh, what a pity. I hope you get it!

**Comment by Swapnesh-SerpentCS** [comment]
@YogeshMahera-SerpentCS Can you please Approve so I get an Approval from OCA Bot ?

**Review by YogeshMahera-SerpentCS — APPROVED** [review]

File: web_widget_prefixed_url/__manifest__.py
Comment (tarteo) [inline web_widget_prefixed_url/__manifest__.py]: Change "https://github.com/OCA/Web" to  "https://github.com/OCA/web" (lower case)

File: web_widget_prefixed_url/README.rst
Comment (tarteo) [inline web_widget_prefixed_url/README.rst]: Change 'prefixed_url:' to 'prefixed_url' or can you change it to the module name?
I prefer using the readme generator but my understanding is that it's not required.

File: web_widget_prefixed_url/README.rst
Comment (tarteo) [inline web_widget_prefixed_url/README.rst]: Use a code block here.

**Review by tarteo — CHANGES_REQUESTED** [review]

**Review by Tardo — APPROVED** [review]
LGTM :+1:

**Review by daramousk — CHANGES_REQUESTED** [review]
Hello and thanks for this contribution,

I am not sure I understand the use cases of this module. If we have a url, we can already use `UrlWidget`.

File: web_widget_prefixed_url/README.rst
Comment (Swapnesh-SerpentCS) [reply web_widget_prefixed_url/README.rst]: Good Catch :+1: 

I changed it to `prefixed_url`

**Comment by Swapnesh-SerpentCS** [comment]
@tarteo  Suggested changes done.



@daramousk 
> If we have a url, we can already use UrlWidget.

We can adapt these changes over there as well. (Maybe from 12.0 )
Have a look on Conversation after https://github.com/OCA/web/pull/1234#issuecomment-511679776

**Comment by Swapnesh-SerpentCS** [comment]
@tarteo  @pedrobaeza  Can we merge this now?

**Comment by pedrobaeza** [comment]
@tarteo should dismiss or approve his review for continuing

**Review by tarteo — APPROVED** [review]

**Comment by Swapnesh-SerpentCS** [comment]
Finally :pray:

**Pull request merged**
